src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -428,6 +474,63 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
     }
 }
 
+template <typename HWIntrinsicSwitchCaseBody>
+void CodeGen::genJumpTableFallback(unsigned                  maxImmValue,
+                                   regNumber                 nonConstImmReg,
+                                   regNumber                 baseReg,
+                                   regNumber                 offsReg,
+                                   HWIntrinsicSwitchCaseBody emitSwCase)
+{
+    assert(nonConstImmReg != REG_NA);
+    emitter* emit = getEmitter();
+
+    const int   maxByte = 256;
+    BasicBlock* jmpTable[maxByte];
+
+    unsigned jmpTableBase = emit->emitBBTableDataGenBeg(maxImmValue, true);
+    unsigned jmpTableOffs = 0;
+
+    // Emit the jump table
+    for (unsigned i = 0; i < maxImmValue; i++)
+    {
+        jmpTable[i] = genCreateTempLabel();
+        JITDUMP("            DD      L_M%03u_BB%02u\n", Compiler::s_compMethodsCount, jmpTable[i]->bbNum);
+        emit->emitDataGenData(i, jmpTable[i]);
+    }
+
+    emit->emitDataGenEnd();
+
+    if (maxImmValue < maxByte - 1)
+    {
+        emit->emitIns_R_I(INS_cmp, EA_4BYTE, nonConstImmReg, maxImmValue);
+        emitJumpKind jge = genJumpKindForOper(GT_GE, CK_UNSIGNED);
+        genJumpToThrowHlpBlk(jge, SCK_ARG_RNG_EXCPN);
```

Comment:
The runtime intermittently gets "failed to find exception throw block". @CarolEidt @tannergooding @mikedn @jkotas  Do you know why?

```
19:06:15         CLRJIT! CHECK::Trigger + 0x30E (0x7354b789)
19:06:15         CLRJIT! CodeGen::genJumpToThrowHlpBlk + 0x13C (0x73435bd7)
19:06:15         CLRJIT! CodeGen::genJumpTableFallback<<lambda_8c168ac1d3dcdd43425f08b8cdfa80f0> > + 0xF7 (0x7353e539)
19:06:15         CLRJIT! CodeGen::genHWIntrinsic + 0x2AE (0x7353ec79)
19:06:15         CLRJIT! CodeGen::genCodeForTreeNode + 0x43D (0x73516b55)
19:06:15         CLRJIT! CodeGen::genCodeForBBlist + 0x80E (0x7343936e)
19:06:15         CLRJIT! CodeGen::genGenerateCode + 0x283 (0x73434f23)
19:06:15         CLRJIT! Compiler::compCompile + 0x72D (0x73440571)
19:06:15         CLRJIT! Compiler::compCompileHelper + 0x589 (0x73441767)
19:06:15         CLRJIT! Compiler::compCompile + 0x622 (0x73440c53)
19:06:15             File: d:\j\workspace\x86_checked_w---b1af1919\src\jit\codegencommon.cpp Line: 2826
19:06:15             Image: D:\j\workspace\x86_checked_w---b1af1919\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\CoreRun.exe
```

```cpp
        {
            /* Find the helper-block which raises the exception. */

            Compiler::AddCodeDsc* add =
                compiler->fgFindExcptnTarget(codeKind, compiler->bbThrowIndex(compiler->compCurBB));
            PREFIX_ASSUME_MSG((add != nullptr), ("ERROR: failed to find exception throw block"));
            tgtBlk = add->acdDstBlk;
        }
```

Reply:
Range checks must be exposed in the importer, so that they can be appropriately modeled in the control flow. Check out the places in simd.cpp where it generates `GT_SIMD_CHK`.

Reply:
@CarolEidt Thanks for pointing out. IMO, I still do not like the current solution to non-const (that hardcodes a jump-table fallback in the compiler), which
- needs more IR to represent the range-check because the out-of-range constant argument will become a variable in the recursive function-body.
- the codegen always needs to consider keeping consistent semantics with the front-end. 

So I think the software switch-table fallback can significantly simplify the compiler implementation that only deals with constant imm-argument. 

cc @tannergooding

Reply:
I thought we determined that, we could also do this as:
```C#
result Intrinsic(op1, op2)
{
    RangeCheck();
    return Intrinsic(op1, op2);
}
```

Which only leaves the switch table up to the JIT.

Reply:
If we do it all in the JIT however, it still seems like a one time cost in implementing a helper function to do this, and it only needs to happen in the fallback case, which is already on its own codepath.

Reply:
I think the software switch-table is not a problem now. In the future, if the compiler determines to optimize away, we can use an attribute to disable the optimization on these intrinsics.

Reply:
> In the future, if the compiler determines to optimize away the switch, we can use an attribute to disable the optimization on these intrinsics.

There are multiple different compilers, and some (such as the C# compilers) don't have an option to disable optimizations for a given method. This would only work for an AOT or JIT compiler that was performing the optimization, and only if it respected the flag.

I think it would be good to better understand the specific pushback here. We already discussed this at some length in at least one of the other threads and landed on one of these two approaches:
* do the range check in managed and emit the switch table in the JIT
* emit the range check and the switch table in the JIT

Reply:
As far as I'm concerned, it is a non-starter to consider introducing a throw in codegen that has not been introduced into the IR earlier. In addition to the technical infeasibility, it would render it impossible to apply range-check optimizations to eliminate it.

I am mildly in favor of emitting both the range check and the switch table in the JIT, primarily because it doesn't rely on the IL implementation and the JIT to both (and separately) specify the appropriate range values.